File: src/main/java/org/kite9/diagram/visualization/planarization/rhd/layout/AbstractTopDownLayoutStrategy.java
package org.kite9.diagram.visualization.planarization.rhd.layout;

import org.kite9.diagram.common.elements.RoutingInfo;
import org.kite9.diagram.common.hints.PositioningHints;
import org.kite9.diagram.model.position.Direction;
import org.kite9.diagram.model.position.Layout;
import org.kite9.diagram.visualization.planarization.rhd.GroupAxis;
import org.kite9.diagram.visualization.planarization.rhd.GroupPhase;
import org.kite9.diagram.visualization.planarization.rhd.GroupPhase.CompoundGroup;
import org.kite9.diagram.visualization.planarization.rhd.GroupPhase.Group;
import org.kite9.diagram.visualization.planarization.rhd.grouping.GroupResult;
import org.kite9.diagram.visualization.planarization.rhd.grouping.directed.DirectedLinkManager;
import org.kite9.diagram.visualization.planarization.rhd.position.RoutableHandler2D;
import org.kite9.framework.logging.Kite9Log;
import org.kite9.framework.logging.Logable;
import org.kite9.framework.logging.LogicException;

/**
 * Provides the basic code for a top-down approach to laying out groups, but doesn't specify 
 * either the algorithm used to choose the approach or the ordering of the groups.
 * 
 * @author robmoffat
 *
 */
public abstract class AbstractTopDownLayoutStrategy implements LayoutStrategy, Logable {
	
	private static final double TOLERANCE = 0.0000001;

	
	public AbstractTopDownLayoutStrategy(RoutableHandler2D rh) {
		this.rh = rh;
	}
	
	Kite9Log log = new Kite9Log(this);
	
	protected RoutableHandler2D rh;

	private void chooseBestCompoundGroupPlacement(GroupPhase gp, GroupPhase.CompoundGroup gg) {
		rh.clearTempPositions(true);
		rh.clearTempPositions(false);
		GroupAxis gt = gg.getAxis();
		Layout ld = gg.getLayout();
		boolean canBeHoriz = gt.isHorizontal() && gt.isLayoutRequired();
		boolean canBeVert = gt.isVertical() && gt.isLayoutRequired();
		boolean horizLayoutUnknown = ((ld == null) || (ld == Layout.HORIZONTAL)) && canBeHoriz;
		boolean vertLayoutUnknown = ((ld == null) || (ld == Layout.VERTICAL)) && canBeVert;
		boolean canDecideLayout = (horizLayoutUnknown || vertLayoutUnknown);
		log.send("Group A: "+gg.getA());
		log.send("Group B: "+gg.getB());
		
		if (canDecideLayout) {
			boolean layoutNeeded = groupsNeedLayout(gg.getA(), gg.getB(), horizLayoutUnknown, vertLayoutUnknown, ld);

			if (layoutNeeded) {
				// this is the expensive part - layout is required. Choose one
				Layout hintedLayout = getHintedLayout(gg, canBeHoriz, canBeVert, ld);
				PlacementApproach best = null;
				best = tryPlacement(gp, gg, best, filterLayout(hintedLayout, horizLayoutUnknown, vertLayoutUnknown), canBeHoriz, canBeVert, true);
				best = tryPlacement(gp, gg, best, filterLayout(Layout.reverse(hintedLayout), horizLayoutUnknown, vertLayoutUnknown), canBeHoriz, canBeVert, false);
				best = tryPlacement(gp, gg, best, filterLayout(Layout.rotateClockwise(hintedLayout), horizLayoutUnknown, vertLayoutUnknown), canBeHoriz, canBeVert, false);
				best = tryPlacement(gp, gg, best, filterLayout(Layout.rotateAntiClockwise(hintedLayout), horizLayoutUnknown, vertLayoutUnknown), canBeHoriz, canBeVert, false);
	
				if (best != null) {
					best.choose();
				}
			} 
		
		} else { 
			PlacementApproach pa = createPlacementApproach(gp, gg, ld, canBeHoriz, canBeVert, true);
			pa.choose();
			log.send("Group layout = "+ld);
		}
	}

	private Layout filterLayout(Layout naturalLayout, boolean horiz, boolean vert) {
		switch (naturalLayout) {
		case LEFT:
		case RIGHT:
			return horiz ? naturalLayout : null;
		case DOWN:
		case UP:
			return vert ? naturalLayout : null;
		default:
			throw new LogicException("Layout should be definite for an approach: "+naturalLayout);
		}
	}
	
	/**
	 * Layout can be hinted either by the {@link PositioningHints} of the groups, or by the ordinal
	 * order of the elements in the container.  If the layout of the container is HORIZONTAL or
	 * VERTICAL, favour the ordinal, otherwise favour the {@link PositioningHints} where available.
	 */
	private Layout getHintedLayout(GroupPhase.CompoundGroup gg, boolean setHoriz, boolean setVert, Layout prescribed) {
		Integer bx = null;
		Integer by = null;
		Layout out = null;
		Group a = gg.getA();
		Group b = gg.getB();
		
		if (setVert) {
			if (prescribed == Layout.VERTICAL) {
				return getVerticalOrdinalLayout(a, b);
			} else if (prescribed == null) {
				by = PositioningHints.compareEitherYBounds(a.getHints(), b.getHints());
				out = getVerticalOrdinalLayout(a, b);
			}
		}
		
		if (setHoriz) {
			if (prescribed == Layout.HORIZONTAL) {
				return getHorizontalOrdinalLayout(a, b);
			} else if (prescribed == null) {
				bx = PositioningHints.compareEitherXBounds(a.getHints(), b.getHints());
				out = getHorizontalOrdinalLayout(a, b);
			}
		}
		
		if (bx != null) {
			if (1 == bx) {
				return Layout.LEFT;
			}else if (-1 == bx) {
				return Layout.RIGHT;
			}
		}
		
		if (by != null) {
			if (1 == by) {
				return Layout.UP;
			} else if (-1 == by) {
				return Layout.DOWN;
			}
		}
		
		return out;
	}

	private Layout getHorizontalOrdinalLayout(Group a, Group b) {
		return a.getGroupOrdinal() < b.getGroupOrdinal() ? Layout.RIGHT: Layout.LEFT;
	}

	private Layout getVerticalOrdinalLayout(Group a, Group b) {
		return a.getGroupOrdinal() < b.getGroupOrdinal() ? Layout.DOWN: Layout.UP;
	}

	private boolean groupsNeedLayout(Group a, Group b, boolean horizLayoutUnknown, boolean vertLayoutUnknown, Layout l) {
		if (groupsOverlap(a, b)) {
			return true;
		}

		boolean straightVerticals = groupsHaveStraightEdges(a, b, false);
		
		if ((horizLayoutUnknown || (l==Layout.VERTICAL)) && straightVerticals) {
			return true;
		}
		
		boolean straightHorizontals = groupsHaveStraightEdges(a, b, true);
		
		if ((vertLayoutUnknown || (l==Layout.HORIZONTAL)) && straightHorizontals) {
			return true;
		}
		
		return false;
	}

	private boolean groupsHaveStraightEdges(Group a, Group b, boolean horiz) {
		int mask = DirectedLinkManager.createMask(null, false, false, horiz ? Direction.LEFT : Direction.UP, horiz ? Direction.RIGHT : Direction.DOWN);
		boolean aHasLinks = a.getLinkManager().subset(mask).size() > 0;
		boolean bHasLinks = b.getLinkManager().subset(mask).size() > 0;
		return aHasLinks || bHasLinks;
	}

	private boolean groupsOverlap(Group a, Group b) {
		RoutingInfo ari = a.getAxis().getPosition(rh, true);
		RoutingInfo bri = b.getAxis().getPosition(rh, true);
		return rh.overlaps(ari, bri);
	}

	private PlacementApproach tryPlacement(GroupPhase gp, GroupPhase.CompoundGroup gg, PlacementApproach best, Layout d, boolean setHoriz, boolean setVert, boolean natural) {
		if ((d!=null) && ((best == null) || (best.getScore() > 0))) {
			PlacementApproach newpl = createPlacementApproach(gp, gg, d, setHoriz, setVert, natural);
			newpl.evaluate();
			log.send(log.go() ? null : gg.getGroupNumber() + " going " + d + "  score: " + newpl.getScore());
			
			if (best == null) {
				return newpl;
			} else if (best.getScore() <= newpl.getScore() + TOLERANCE) {
				return best;
			} else if (best.getScore() >= newpl.getScore() + TOLERANCE) {
				return newpl;
			} else {
				if (best.isNatural()) {
					return best;
				} else if (newpl.isNatural()) {
					return newpl;
				} else {
					return best;
				}
 			}
		}
		return best;
	}
	
	protected abstract PlacementApproach createPlacementApproach(GroupPhase gp, GroupPhase.CompoundGroup gg, Layout ld,
			boolean setHoriz, boolean setVert, boolean natural);

	private void chooseBestPlacement(GroupPhase gp, LayoutQueue lq) {
		Group g = lq.poll();
		while (g != null) {
			g.getAxis().getPosition(rh, false);
			log.send(log.go() ? null : "Ordering "+g.getGroupNumber()+ " size="+g.getSize()+ " links="+g.getLinkManager().getLinkCount());
			StringBuilder out = new StringBuilder(1000);
			log.send(out.toString());
			if (g instanceof CompoundGroup) {
				CompoundGroup cg = (CompoundGroup) g;
				chooseBestCompoundGroupPlacement(gp, cg);
				lq.complete(cg);
				lq.offer(cg.getA());
				lq.offer(cg.getB());
			}  
			
			g = lq.poll();
		}
	}
	
	
	
	@Override
	public String getPrefix() {
		return "TDLS";
	}

	@Override
	public boolean isLoggingEnabled() {
		return true;
	}

	@Override
	public void layout(GroupPhase gp, GroupResult mr, LayoutQueue lq) {
		Group g = mr.groups().iterator().next();
		lq.offer(g);
		chooseBestPlacement(gp, lq);
		//rh.outputSettings();
	}

}

File: src/main/kotlin/org/kite9/diagram/visualization/planarization/rhd/layout/AbstractTopDownLayoutStrategy.kt
package org.kite9.diagram.visualization.planarization.rhd.layout

import org.kite9.diagram.common.hints.PositioningHints
import org.kite9.diagram.common.hints.PositioningHints.compareEitherXBounds
import org.kite9.diagram.common.hints.PositioningHints.compareEitherYBounds
import org.kite9.diagram.logging.Kite9Log
import org.kite9.diagram.logging.Logable
import org.kite9.diagram.logging.LogicException
import org.kite9.diagram.model.position.Direction
import org.kite9.diagram.model.position.Layout
import org.kite9.diagram.model.position.Layout.Companion.reverse
import org.kite9.diagram.model.position.Layout.Companion.rotateAntiClockwise
import org.kite9.diagram.model.position.Layout.Companion.rotateClockwise
import org.kite9.diagram.visualization.planarization.rhd.grouping.basic.group.Group
import org.kite9.diagram.visualization.planarization.rhd.grouping.GroupResult
import org.kite9.diagram.visualization.planarization.rhd.grouping.basic.group.CompoundGroup
import org.kite9.diagram.visualization.planarization.rhd.grouping.directed.group.DirectedLinkManager
import org.kite9.diagram.visualization.planarization.rhd.position.RoutableHandler2D

/**
 * Provides the basic code for a top-down approach to laying out groups, but doesn't specify
 * either the algorithm used to choose the approach or the ordering of the groups.
 *
 * @author robmoffat
 */
abstract class AbstractTopDownLayoutStrategy(val rh: RoutableHandler2D) : LayoutStrategy, Logable {

    @JvmField
	var log = Kite9Log(this)

    private fun chooseBestCompoundGroupPlacement(gg: CompoundGroup) {
        rh.clearTempPositions(true)
        rh.clearTempPositions(false)
        val gt = gg.axis
        val ld = gg.layout
        val canBeHoriz = gt.isHorizontal && gt.isLayoutRequired
        val canBeVert = gt.isVertical && gt.isLayoutRequired
        val horizLayoutUnknown = (ld == null || ld === Layout.HORIZONTAL) && canBeHoriz
        val vertLayoutUnknown = (ld == null || ld === Layout.VERTICAL) && canBeVert
        val canDecideLayout = horizLayoutUnknown || vertLayoutUnknown
        log.send("Group A: " + gg.a)
        log.send("Group B: " + gg.b)
        if (canDecideLayout) {
            val layoutNeeded = groupsNeedLayout(gg.a, gg.b, horizLayoutUnknown, vertLayoutUnknown, ld)
            if (layoutNeeded) {
                // this is the expensive part - layout is required. Choose one
                val hintedLayout = getHintedLayout(gg, canBeHoriz, canBeVert, ld)
                var best: PlacementApproach? = null
                best = tryPlacement(
                    gg,
                    best,
                    filterLayout(hintedLayout, horizLayoutUnknown, vertLayoutUnknown),
                    canBeHoriz,
                    canBeVert,
                    true
                )
                best = tryPlacement(
                    gg,
                    best,
                    filterLayout(reverse(hintedLayout), horizLayoutUnknown, vertLayoutUnknown),
                    canBeHoriz,
                    canBeVert,
                    false
                )
                best = tryPlacement(
                    gg,
                    best,
                    filterLayout(rotateClockwise(hintedLayout), horizLayoutUnknown, vertLayoutUnknown),
                    canBeHoriz,
                    canBeVert,
                    false
                )
                best = tryPlacement(
                    gg,
                    best,
                    filterLayout(rotateAntiClockwise(hintedLayout), horizLayoutUnknown, vertLayoutUnknown),
                    canBeHoriz,
                    canBeVert,
                    false
                )
                best?.choose()
            }
        } else {
            val pa = createPlacementApproach(gg, ld, canBeHoriz, canBeVert, true)
            pa.choose()
            log.send("Group layout = $ld")
        }
    }

    private fun filterLayout(naturalLayout: Layout?, horiz: Boolean, vert: Boolean): Layout? {
        return when (naturalLayout) {
            Layout.LEFT, Layout.RIGHT -> if (horiz) naturalLayout else null
            Layout.DOWN, Layout.UP -> if (vert) naturalLayout else null
            else -> throw LogicException("Layout should be definite for an approach: $naturalLayout")
        }
    }

    /**
     * Layout can be hinted either by the [PositioningHints] of the groups, or by the ordinal
     * order of the elements in the container.  If the layout of the container is HORIZONTAL or
     * VERTICAL, favour the ordinal, otherwise favour the [PositioningHints] where available.
     */
    private fun getHintedLayout(gg: CompoundGroup, setHoriz: Boolean, setVert: Boolean, prescribed: Layout?): Layout? {
        var bx: Int? = null
        var by: Int? = null
        var out: Layout? = null
        val a = gg.a
        val b = gg.b
        if (setVert) {
            if (prescribed === Layout.VERTICAL) {
                return getVerticalOrdinalLayout(a, b)
            } else if (prescribed == null) {
                by = compareEitherYBounds(a.hints, b.hints)
                out = getVerticalOrdinalLayout(a, b)
            }
        }
        if (setHoriz) {
            if (prescribed === Layout.HORIZONTAL) {
                return getHorizontalOrdinalLayout(a, b)
            } else if (prescribed == null) {
                bx = compareEitherXBounds(a.hints, b.hints)
                out = getHorizontalOrdinalLayout(a, b)
            }
        }
        if (bx != null) {
            if (1 == bx) {
                return Layout.LEFT
            } else if (-1 == bx) {
                return Layout.RIGHT
            }
        }
        if (by != null) {
            if (1 == by) {
                return Layout.UP
            } else if (-1 == by) {
                return Layout.DOWN
            }
        }
        return out
    }

    private fun getHorizontalOrdinalLayout(a: Group, b: Group): Layout {
        return if (a.groupOrdinal < b.groupOrdinal) Layout.RIGHT else Layout.LEFT
    }

    private fun getVerticalOrdinalLayout(a: Group, b: Group): Layout {
        return if (a.groupOrdinal < b.groupOrdinal) Layout.DOWN else Layout.UP
    }

    private fun groupsNeedLayout(
        a: Group,
        b: Group,
        horizLayoutUnknown: Boolean,
        vertLayoutUnknown: Boolean,
        l: Layout?
    ): Boolean {
        if (groupsOverlap(a, b)) {
            return true
        }
        val straightVerticals = groupsHaveStraightEdges(a, b, false)
        if ((horizLayoutUnknown || l === Layout.VERTICAL) && straightVerticals) {
            return true
        }
        val straightHorizontals = groupsHaveStraightEdges(a, b, true)
        return if ((vertLayoutUnknown || l === Layout.HORIZONTAL) && straightHorizontals) {
            true
        } else false
    }

    private fun groupsHaveStraightEdges(a: Group, b: Group, horiz: Boolean): Boolean {
        val mask = DirectedLinkManager.createMask(
            null,
            false,
            false,
            if (horiz) Direction.LEFT else Direction.UP,
            if (horiz) Direction.RIGHT else Direction.DOWN
        )
        val aHasLinks = a.linkManager.subset(mask).size > 0
        val bHasLinks = b.linkManager.subset(mask).size > 0
        return aHasLinks || bHasLinks
    }

    private fun groupsOverlap(a: Group, b: Group): Boolean {
        val ari = a.axis.getPosition(rh, true)
        val bri = b.axis.getPosition(rh, true)
        return rh.overlaps(ari, bri)
    }

    private fun tryPlacement(
        gg: CompoundGroup,
        best: PlacementApproach?,
        d: Layout?,
        setHoriz: Boolean,
        setVert: Boolean,
        natural: Boolean
    ): PlacementApproach? {
        if (d != null && (best == null || best.score > 0)) {
            val newpl = createPlacementApproach(gg, d, setHoriz, setVert, natural)
            newpl.evaluate()
            log.send(if (log.go()) null else "${gg.groupNumber} going $d  score: ${newpl.score}")
            return if (best == null) {
                newpl
            } else if (best.score <= newpl.score + TOLERANCE) {
                best
            } else if (best.score >= newpl.score + TOLERANCE) {
                newpl
            } else {
                if (best.natural) {
                    best
                } else if (newpl.natural) {
                    newpl
                } else {
                    best
                }
            }
        }
        return best
    }

    protected abstract fun createPlacementApproach(
        gg: CompoundGroup, ld: Layout?,
        setHoriz: Boolean, setVert: Boolean, natural: Boolean
    ): PlacementApproach

    private fun chooseBestPlacement(lq: LayoutQueue) {
        var g = lq.poll()
        while (g != null) {
            g.axis.getPosition(rh, false)
            log.send(if (log.go()) null else "Ordering " + g.groupNumber + " size=" + g.size + " links=" + g.linkManager.linkCount)
            val out = StringBuilder(1000)
            log.send(out.toString())
            if (g is CompoundGroup) {
                val cg = g
                chooseBestCompoundGroupPlacement(cg)
                lq.complete(cg)
                lq.offer(cg.a)
                lq.offer(cg.b)
            }
            g = lq.poll()
        }
    }

    override val prefix: String
        get() = "TDLS"
    override val isLoggingEnabled: Boolean
        get() = true

    override fun layout(mr: GroupResult, lq: LayoutQueue) {
        val g = mr.groups().iterator().next()
        lq.offer(g)
        chooseBestPlacement(lq)
    }

    companion object {
        private const val TOLERANCE = 0.0000001
    }
}